x-pack/plugins/reporting/server/lib/content_stream.ts

```diff
@@ -45,55 +171,152 @@ export class ContentStream extends Duplex {
+    const hits = response?.body.hits?.hits?.[0];
+
+    return hits?._source?.output.content;
+  }
+
+  private isRead() {
+    return this.jobSize != null && this.bytesRead >= this.jobSize;
+  }
+
+  async _read() {
+    try {
+      const content = this.chunksRead ? await this.readChunk() : await this.readHead();
+      if (!content) {
+        this.push(null);
+        return;
       }
 
-      this.push(null);
+      const buffer = await this.decode(content);
+
+      this.push(buffer);
+      this.chunksRead++;
+      this.bytesRead += buffer.byteLength;
+
+      if (this.isRead()) {
+        this.push(null);
+      }
     } catch (error) {
       this.destroy(error);
     }
   }
 
-  _write(chunk: Buffer | string, _encoding: string, callback: Callback) {
-    this.buffer += typeof chunk === 'string' ? chunk : chunk.toString();
-    callback();
+  private async removeChunks() {
+    const { id, index } = this.document;
+
+    await this.client.deleteByQuery({
+      index,
+      body: {
+        query: {
+          match: { parent_id: id },
+        },
+      },
+    });
   }
 
-  async _final(callback: Callback) {
-    try {
-      const { body } = await this.client.update<ReportDocument>({
-        ...this.document,
-        body: {
-          doc: {
-            output: {
-              content: this.buffer,
-            },
-          },
+  private async writeHead(content: string) {
+    const { body } = await this.client.update<ReportSource>({
+      ...this.document,
+      body: {
+        doc: {
+          output: { content },
         },
-      });
+      },
+    });
+
+    ({ _primary_term: this.primaryTerm, _seq_no: this.seqNo } = body);
+  }
+
+  private async writeChunk(content: string) {
+    const { id: parentId, index } = this.document;
+    const id = this.puid.generate();
+
+    await this.client.index<ChunkSource>({
+      id,
+      index,
+      body: {
+        parent_id: parentId,
+        output: {
+          content,
+          chunk: this.chunksWritten,
+        },
+      },
+    });
+  }
+
+  private async flush(size = this.buffer.byteLength) {
+    const chunk = this.buffer.slice(0, size);
+    if (!chunk.byteLength) {
+      return;
+    }
+
+    const content = await this.encode(chunk);
+
+    if (!this.chunksWritten) {
+      await this.removeChunks();
+      await this.writeHead(content);
+    } else {
+      await this.writeChunk(content);
+    }
+
+    this.chunksWritten++;
+    this.bytesWritten += chunk.byteLength;
+    this.buffer = this.buffer.slice(size);
+  }
+
+  async _write(chunk: Buffer | string, encoding: BufferEncoding, callback: Callback) {
+    this.buffer = Buffer.concat([
+      this.buffer,
+      Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding),
+    ]);
+
+    try {
+      const maxChunkSize = await this.getMaxChunkSize();
+
+      while (this.buffer.byteLength >= maxChunkSize) {
+        await this.flush(maxChunkSize);
```

Comment:
Can we add a debug line that prints out every time a chunk of output gets flushed?

Reply:
I have added more extensive logging that can help in the case of debugging. Thanks for pointing this out 👍